homeassistant/components/songpal/config_flow.py

```diff
@@ -0,0 +1,141 @@
+"""Config flow to configure songpal component."""
+import logging
+from typing import Optional
+from urllib.parse import urlparse
+
+from songpal import Device, SongpalException
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components import ssdp
+from homeassistant.const import CONF_HOST, CONF_NAME
+
+from .const import CONF_ENDPOINT, DOMAIN  # pylint: disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SongpalConfig:
+    """Device Configuration."""
+
+    def __init__(self, name, host, endpoint):
+        """Initialize Configuration."""
+        self.name = name
+        self.host = host
+        self.endpoint = endpoint
+
+
+class SongpalConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Songpal configuration flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    def __init__(self):
+        """Initialize the flow."""
+        self.conf: Optional[SongpalConfig] = None
+
+    async def _show_setup_form(self, user_input=None, errors=None):
```

Comment:
I don't think it makes sense having this separate function since it's only used for the user step. Better to have this logic in `async_step_user` since that is what you are doing in `init`

Reply:
@shenxn so sorry, I didn't realize you used this function call twice, so utilizing a function makes sense here if you want to revert, I would just update the function name to make it clear that this function is for the `user` step